x-pack/plugins/infra/public/components/logging/log_text_stream/log_entry_row.tsx

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+// import { darken, transparentize } from 'polished';
+import React, { useState, useCallback, useMemo } from 'react';
+
+import euiStyled from '../../../../../../common/eui_styled_components';
+import {
+  LogEntry,
+  isFieldColumn,
+  isMessageColumn,
+  isTimestampColumn,
+} from '../../../utils/log_entry';
+import {
+  LogColumnConfiguration,
+  isTimestampLogColumnConfiguration,
+  isMessageLogColumnConfiguration,
+  isFieldLogColumnConfiguration,
+} from '../../../utils/source_configuration';
+import { TextScale } from '../../../../common/log_text_scale';
+import { LogEntryColumn, LogEntryColumnWidth } from './log_entry_column';
+import { LogEntryFieldColumn } from './log_entry_field_column';
+import { LogEntryDetailsIconColumn } from './log_entry_icon_column';
+import { LogEntryMessageColumn } from './log_entry_message_column';
+import { LogEntryTimestampColumn } from './log_entry_timestamp_column';
+import { monospaceTextStyle } from './text_styles';
+
+interface LogEntryRowProps {
+  boundingBoxRef?: React.Ref<Element>;
+  columnConfigurations: LogColumnConfiguration[];
+  columnWidths: LogEntryColumnWidth[];
+  isHighlighted: boolean;
+  logEntry: LogEntry;
+  openFlyoutWithItem: (id: string) => void;
+  scale: TextScale;
+  wrap: boolean;
+}
+
+export const LogEntryRow = ({
+  boundingBoxRef,
+  columnConfigurations,
+  columnWidths,
+  isHighlighted,
+  logEntry,
+  openFlyoutWithItem,
+  scale,
+  wrap,
+}: LogEntryRowProps) => {
+  const [isHovered, setIsHovered] = useState(false);
+
+  const setItemIsHovered = useCallback(() => {
+    setIsHovered(true);
+  }, []);
+
+  const setItemIsNotHovered = useCallback(() => {
+    setIsHovered(false);
+  }, []);
+
+  const openFlyout = useCallback(() => openFlyoutWithItem(logEntry.gid), [
+    openFlyoutWithItem,
+    logEntry.gid,
+  ]);
+
+  const iconColumnWidth = useMemo(() => columnWidths[columnWidths.length - 1], [columnWidths]);
+
+  return (
+    <LogEntryRowWrapper
+      data-test-subj="streamEntry logTextStreamEntry"
+      innerRef={
+        /* Workaround for missing RefObject support in styled-components */
+        boundingBoxRef as any
+      }
+      onMouseEnter={setItemIsHovered}
+      onMouseLeave={setItemIsNotHovered}
+      scale={scale}
+    >
+      {logEntry.columns.map((column, columnIndex) => {
+        const columnConfiguration = columnConfigurations[columnIndex];
+        const columnWidth = columnWidths[columnIndex];
+
+        if (isTimestampColumn(column) && isTimestampLogColumnConfiguration(columnConfiguration)) {
```

Comment:
Are there scenarios where these would fall out of sync? I.e. the `column` **is** a timestamp column, but the columnConfiguration **isn't** a timestamp column configuration. I know these are mapped by index position, rather than something like ID, so just wondering.

Reply:
Good question! Yes, I've seen these situations. Since the data are loaded asynchronously after a configuration change, the column configuration might be newer than the loaded data. If React renders "in between", they will not match.

I have considered incorporating the column uuid in the returned data, but since it would have required an API change I went with index-based correlation here.

Apart from that, these `isTimestamp*` functions also act as type guards so we can benefit from narrower types in the condition branch body, i.e. when accessing `columnConfiguration.timestampColumn.id` or `column.timestamp`.